### PR TITLE
Added hash functions to Command, State and Configuration (fixes #109)

### LINKED ISSUE #109: Implement __hash__ for Command, State and Configuration

### PR DESCRIPTION
**Pull request by ChrisTimperley**

### CODE REVIEW

**Comment by afsafzal** [comment]
@ChrisTimperley Could you please take a look at this too?

File: houston/command.py [hunk 1]
```diff
@@ -222,6 +222,11 @@ def __getitem__(self, name: str) -> Any:
             raise KeyError(msg)
         return getattr(self, param._field)
 
+    def __hash__(self) -> int:
+        all_params = [self[p.name] for p in self.__class__.parameters]
+        all_params.insert(0, self.uid)
```
Comment (ChrisTimperley) [inline houston/command.py]: Maybe `(self.uid,) + (self[p.name] for p in self.__class__.parameters)`?

File: houston/configuration.py [hunk 2]
```diff
@@ -111,13 +111,19 @@ def __getitem__(self, name: str) -> Any:
...
+        all_opts = [self[opt.name] for opt in self.__class__.options]
+        all_opts.insert(0, self.__class__.__name__)
+        return hash(tuple(all_opts))
 
     def __eq__(self, other: 'Configuration') -> bool:
         if type(self) != type(other):
-            msg = "illegal comparison of configurations: [{}] vs. [{}]"
-            msg = msg.format(self.__class__.__name__, state.__class__.__name__)
-            raise Exception(msg)  # FIXME use HoustonException
+            msg = "comparison of different classesof configs: [{}] vs. [{}]"
```
Comment (ChrisTimperley) [inline houston/configuration.py]: Typo in string

File: houston/configuration.py [hunk 3]
```diff
@@ -111,13 +111,19 @@ def __getitem__(self, name: str) -> Any:
...
+            msg = msg.format(self.__class__.__name__, other.__class__.__name__)
+            logger.warning(msg)
```
Comment (ChrisTimperley) [inline houston/configuration.py]: Raise an actual warning via `warnings.warn`. The logger will automatically capture it.

File: houston/environment.py [hunk 4]
```diff
@@ -34,5 +34,13 @@ def __getitem__(self, name: str):
         """
         return self.__values[name]
 
+    def __eq__(self, other: 'Environment') -> bool:
+        return self.to_json() == other.to_json()
```
Comment (ChrisTimperley) [inline houston/environment.py]: Using `to_json` to compare is a bad practice that just happens to work in this instance. I don't think we need to change anything for now, but we'll want to keep a note of this for later. Ideally, Environment should be reimplemented in the same manner as Command and State.

File: houston/command.py (hunk 1 above)
Comment (afsafzal) [reply houston/command.py]: `(self[p.name] for p in self.__class__.parameters)` gives me a `generator` but `tuple(self[p.name] for p in self.__class__.parameters)` should be fine.

File: houston/environment.py [hunk 5]
```diff
@@ -34,5 +34,13 @@ def __getitem__(self, name: str):
...
+
+    def __hash__(self) -> int:
+        vals = list(self.__values.copy().items())
```
Comment (ChrisTimperley) [inline houston/environment.py]: No need to copy. Also, `vals` could just be a `tuple` here?

Comment (ChrisTimperley) [reply houston/environment.py]:
```
vals = tuple(self.__values.items())
vals = sorted(vals, key=lambda x: x[0])
return hash(vals)
```

File: houston/state.py [hunk 6]
```diff
@@ -232,6 +232,11 @@ def exact(self, other: 'State') -> bool:
 
     __eq__ = exact
 
+    def __hash__(self) -> int:
+        all_vars = [self[v.name] for v in self.__class__.variables]
```
Comment (ChrisTimperley) [inline houston/state.py]: Same trick as above: `(self.time_offset,) + (...)`

File: houston/command.py (hunk 1 above)
Comment (ChrisTimperley) [reply houston/command.py]: Ah, that makes sense. Sure.

**Comment by afsafzal** [comment]
> How come there's an exception now?

Because there's exception for all the others and I wanted to be consistent. We can change it if we found a usecase for it.

**Pull request merged**